app/uk/gov/hmrc/customs/declarations/stub/connector/NotificationConnector.scala

```diff
@@ -87,26 +88,44 @@ class NotificationConnector @Inject()(
+      val secondCode = random.nextInt(8999999) + 1000000
+      val witchoutCheck = s"${year}${country}${code}${charSection}${secondCode}"
+      val check = witchoutCheck.zipWithIndex.foldLeft(0) {
+        case (input, (char, index)) => input + (NotificationGenerator.characterValue(char) * (1 << index))
+      }
+      val controlDigit = ((check % 11) % 10).toString
+      witchoutCheck + controlDigit
+    }
+  }
+
+  private def generate(lrn: String, default: String, operation: String, maybeMrn: Option[String]): (FiniteDuration, String) = {
     val delay = extractDelay(lrn)
+    val mrn = defineMrn(maybeMrn)
 
     if (operation == "cancel") {
       lrn.charAt(2) match {
-        case 'S' => (delay, generator.generate(lrn, Seq(CustomsPositionGranted)).toString)
-        case 'D' => (delay, generator.generate(lrn, Seq(CustomsPositionDenied)).toString)
-        case _   => (delay, generator.generate(lrn, Seq(QueryNotificationMessage)).toString)
+        case 'S' => (delay, generator.generate(lrn, mrn, Seq(CustomsPositionGranted)).toString)
+        case 'D' => (delay, generator.generate(lrn, mrn, Seq(CustomsPositionDenied)).toString)
+        case _   => (delay, generator.generate(lrn, mrn, Seq(QueryNotificationMessage)).toString)
       }
     } else {
       lrn.headOption match {
-        case Some('B') => (delay, generator.generate(lrn, List(Rejected)).toString)
-        case Some('C') => (delay, generator.generate(lrn, List(Accepted, Cleared)).toString)
-        case Some('D') => (delay, generator.generate(lrn, List(Accepted, AdditionalDocumentsRequired)).toString)
-        case Some('G') => (delay, generator.generate(lrn, List(Accepted)).toString)
-        case Some('Q') => (delay, generator.generate(lrn, List(QueryNotificationMessage)).toString)
-        case Some('R') => (delay, generator.generate(lrn, List(Received)).toString)
-        case Some('U') => (delay, generator.generate(lrn, List(UndergoingPhysicalCheck)).toString)
-        case Some('X') => (delay, generator.generate(lrn, List(GoodsHaveExitedTheCommunity)).toString)
-        case _         => (delay, importsSpecificErrors(lrn, default))
+        case Some('B') => (delay, generator.generate(lrn, mrn, List(Rejected)).toString)
+        case Some('C') => (delay, generator.generate(lrn, mrn, List(Accepted, Cleared)).toString)
+        case Some('D') => (delay, generator.generate(lrn, mrn, List(Accepted, AdditionalDocumentsRequired)).toString)
+        case Some('G') => (delay, generator.generate(lrn, mrn, List(Accepted)).toString)
+        case Some('Q') => (delay, generator.generate(lrn, mrn, List(QueryNotificationMessage)).toString)
+        case Some('R') => (delay, generator.generate(lrn, mrn, List(Received)).toString)
+        case Some('U') => (delay, generator.generate(lrn, mrn, List(UndergoingPhysicalCheck)).toString)
+        case Some('X') => (delay, generator.generate(lrn, mrn, List(GoodsHaveExitedTheCommunity)).toString)
```

Comment:
Small U2Y improvement... method `generate` could return `String` instead of `Elem` in order to avoid to always call `toString` on every resulting XML.

In `generate`...
```
    (<urn:MetaData
      ...
     </urn:MetaData>).toString
```